Preventing users from buying already gifted subscriber items

Hus274 opened the pull request:
closes #7712 
### Changes

Adding a check of the users unopened subscriber gifts when finding the subscriber items that the user doesn't have.

---

UUID: 2a4d30ee-4833-420e-be7c-ad6a6e261c5f

crookedneighbor commented on common/script/content/index.js:
```diff
@@ -38,9 +38,11 @@ _.each(api.mystery, function(v, k) {
+  var mystery = api.mystery;
   var ownedKeys;
   ownedKeys = _.keys((typeof owned.toObject === "function" ? owned.toObject() : void 0) || owned);
+  ownedKeys = _.union( ownedKeys, unopenedGifts);
```
Extra space

Hus274 replied:
fixed

crookedneighbor commented:
This behavior needs test coverage.

khipkin commented on common/script/content/index.js:
```diff
@@ -38,9 +38,11 @@ _.each(api.mystery, function(v, k) {
 
-api.timeTravelerStore = function(owned) {
+api.timeTravelerStore = function(owned, unopenedGifts) {
+  var mystery = api.mystery;
```
this value is not used

Hus274 replied:
fixed

crookedneighbor commented:
Hmmm... there's some general refactoring I want to do on these functions, so I'm going to take this over. Good work on getting it started. Looks like it solves the issue.

Hus274 commented:
Ok sounds good!

crookedneighbor commented on common/script/content/index.js:
```diff
@@ -38,9 +38,11 @@ _.each(api.mystery, function(v, k) {
 });
 
-api.timeTravelerStore = function(owned) {
+api.timeTravelerStore = function(owned, unopenedGifts) {
```
Would it be a simpler api to just pass in the user here?

Hus274 replied:
yeah that makes sense

crookedneighbor commented:
There are still a few pending comments on here, but it looks mostly 👍  to me.

Hus274 commented:
You mentioned above that you were going to finish this up, would you like me to do it instead?

crookedneighbor commented:
Whoops. Totally forgot. Go ahead and finish it. I'm clearly not doing it :smile:

Hus274 commented:
Haha no worries, I can finish it up!

crookedneighbor commented:
@Hus274 Status on this?

Hus274 commented:
@crookedneighbor sorry I've been busy the last week or so.  Should have some time to finish it up this this week. Will plan for either tomorrow or Tuesday.

crookedneighbor commented:
No worries. Just wanted to make sure you weren't waiting on me for anything.

Commit:
Adding the unopened mystery items to the call to get not obtained subscriber items. closes #7712

Hus274 reviewed (commented):
Just going to add up tests here and then I'll resubmit this.

Hus274 commented:
I've committed the changes that have been requested. I'm not sure where to put a test for the timeTravelerStore method as it's in index.js. Any thoughts?

TheHollidayInn commented:
@Hus274 Is it possible to place this function somewhere else? If it must stay, you can put the test here: https://github.com/HabitRPG/habitica/tree/develop/test/content.

TheHollidayInn commented on website/views/options/inventory/time-travelers.jade:
```diff
@@ -21,7 +21,7 @@
       li.customize-menu.inventory-gear
         menu.pets-menu(label='{{::category.text}}', ng-repeat='category in timeTravelersCategories')
-          div(ng-repeat='item in category.items', ng-if='category.identifier === "pets" || category.identifier === "mounts" ? !user.items[category.identifier][item.key] : !user.items.gear.owned[item.key]')
+          div(ng-repeat='item in category.items', ng-if='category.identifier === "pets" || category.identifier === "mounts" ? !user.items[category.identifier][item.key] : (!user.items.gear.owned[item.key] && user.purchased.plan.mysteryItems.indexOf(item.key) == -1)')
```
We should move this if logic to the controller.

Hus274 committed:
refactoring according to pr

Hus274 committed:
Merge remote-tracking branch 'upstream/develop' into 7712

Hus274 committed:
Refactoring according to pr. moved time-travelers to it's own file and added new tests.

Hus274 commented:
@TheHollidayInn I moved the complicated if logic to the controller and removed the ternary operator for clarity. Definitely looks better. I also moved the time traveler's logic into it's own file and added tests around the behavior. Let me know if there's anything else!
